File: packages/python/minions_prompts/prompt_chain.py
"""
PromptChain — traverses follows relations to reconstruct complete version history.
"""

from __future__ import annotations

from minions import Minion

from .storage import PromptStorage


class PromptChain:
    """Traverses ``follows`` relations to reconstruct the complete prompt lineage.

    Args:
        storage: The storage backend to use for minion and relation retrieval.

    Example::

        storage = InMemoryStorage()
        chain = PromptChain(storage)
        versions = chain.get_version_chain(prompt_id)
        latest = chain.get_latest_version(prompt_id)
    """

    def __init__(self, storage: PromptStorage) -> None:
        self._storage = storage

    def get_version_chain(self, prompt_id: str) -> list[Minion]:
        """Return all versions in the chain, sorted oldest first.

        Traverses ``follows`` relations from the given prompt backwards to the
        root, then collects all forward-linked versions.

        Args:
            prompt_id: The ID of any minion in the version chain.

        Returns:
            List of minions in chronological order (oldest first).
        """
        root = self._find_root(prompt_id)
        visited: set[str] = set()
        chain: list[Minion] = []
        queue = [root.id]
        visited.add(root.id)
        chain.append(root)

        while queue:
            current_id = queue.pop(0)
            following = self._storage.get_relations(target_id=current_id, type="follows")
            for rel in following:
                if rel.source_id not in visited:
                    visited.add(rel.source_id)
                    minion = self._storage.get_minion(rel.source_id)
                    if minion:
                        chain.append(minion)
                        queue.append(rel.source_id)

        from datetime import datetime
        chain.sort(key=lambda m: datetime.fromisoformat(m.created_at))
        return chain

    def get_latest_version(self, prompt_id: str) -> Minion:
        """Return the most recently created leaf version.

        Args:
            prompt_id: The ID of any minion in the chain.

        Returns:
            The latest version minion.

        Raises:
            ValueError: If no version chain is found.
        """
        chain = self.get_version_chain(prompt_id)
        if not chain:
            raise ValueError(f"No version chain found for prompt {prompt_id}")

        chain_ids = {m.id for m in chain}
        leaf_nodes = []
        for minion in chain:
            successors = self._storage.get_relations(target_id=minion.id, type="follows")
            chain_successors = [r for r in successors if r.source_id in chain_ids]
            if not chain_successors:
                leaf_nodes.append(minion)

        from datetime import datetime
        leaf_nodes.sort(key=lambda m: datetime.fromisoformat(m.created_at), reverse=True)
        return leaf_nodes[0]

    def get_version_at_date(self, prompt_id: str, date: "datetime") -> Minion | None:
        """Return the version that was active at a specific date.

        Args:
            prompt_id: The ID of any minion in the chain.
            date: The target datetime.

        Returns:
            The active version at that date, or None if none existed yet.
        """
        from datetime import datetime
        chain = self.get_version_chain(prompt_id)
        target_ts = date.timestamp()

        candidates = [
            m for m in chain
            if datetime.fromisoformat(m.created_at).timestamp() <= target_ts
        ]
        if not candidates:
            return None

        candidates.sort(key=lambda m: datetime.fromisoformat(m.created_at), reverse=True)
        return candidates[0]

    def _find_root(self, start_id: str) -> Minion:
        """Walk backwards via follows to find the chain root."""
        visited: set[str] = set()
        current_id = start_id

        while True:
            if current_id in visited:
                raise ValueError(f"Cycle detected in follows chain at {current_id}")
            visited.add(current_id)

            follows_rels = self._storage.get_relations(source_id=current_id, type="follows")
            if not follows_rels:
                minion = self._storage.get_minion(current_id)
                if not minion:
                    raise ValueError(f"Minion not found: {current_id}")
                return minion

            current_id = follows_rels[0].target_id

File: packages/core/src/PromptChain.ts
/**
 * @module minions-prompts/PromptChain
 * Traverses `follows` relations to reconstruct complete prompt version history.
 */

import type { Minion, Relation } from 'minions-sdk';
import type { PromptStorage } from './types.js';

/**
 * Traverses `follows` relations to reconstruct the complete lineage of a prompt.
 *
 * @example
 * ```typescript
 * const storage = new InMemoryStorage();
 * const chain = new PromptChain(storage);
 * const versions = await chain.getVersionChain(promptId);
 * const latest = await chain.getLatestVersion(promptId);
 * ```
 */
export class PromptChain {
  constructor(private readonly storage: PromptStorage) {}

  /**
   * Returns all versions of a prompt in chronological order (oldest first).
   *
   * Traverses `follows` relations from the given prompt/version backwards to
   * the root, then reverses to give oldest-first ordering.
   *
   * @param promptId - The ID of a prompt-template or prompt-version minion.
   * @returns Array of minions forming the version chain, oldest first.
   */
  async getVersionChain(promptId: string): Promise<Minion[]> {
    const visited = new Set<string>();
    const chain: Minion[] = [];

    // BFS forward from root: find all minions that follow (transitively) this root
    // Strategy: start at the given ID and walk backwards via `follows` targets to find root,
    // then walk forwards to collect all versions
    
    // First, find the root by walking backwards
    const root = await this.findRoot(promptId);
    
    // Then collect all versions starting from root using BFS
    const queue = [root.id];
    visited.add(root.id);
    chain.push(root);

    while (queue.length > 0) {
      const currentId = queue.shift()!;
      // Find all minions that follow the current one (i.e., sourceId = X, targetId = currentId, type = follows)
      const followingRelations = await this.storage.getRelations({
        targetId: currentId,
        type: 'follows',
      });

      for (const rel of followingRelations) {
        if (!visited.has(rel.sourceId)) {
          visited.add(rel.sourceId);
          const minion = await this.storage.getMinion(rel.sourceId);
          if (minion) {
            chain.push(minion);
            queue.push(rel.sourceId);
          }
        }
      }
    }

    // Sort by createdAt for consistent ordering
    chain.sort((a, b) => new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime());
    return chain;
  }

  /**
   * Returns the most recent version in the chain.
   *
   * "Latest" is defined as the version with the greatest `createdAt` timestamp
   * among all leaf nodes (versions with no successors).
   *
   * @param promptId - The ID of any minion in the version chain.
   * @returns The latest version minion.
   */
  async getLatestVersion(promptId: string): Promise<Minion> {
    const chain = await this.getVersionChain(promptId);
    if (chain.length === 0) {
      throw new Error(`No version chain found for prompt ${promptId}`);
    }

    // Find leaf nodes (versions with no successors in the chain)
    const chainIds = new Set(chain.map((m) => m.id));
    const leafNodes: Minion[] = [];

    for (const minion of chain) {
      const successors = await this.storage.getRelations({
        targetId: minion.id,
        type: 'follows',
      });
      const chainSuccessors = successors.filter((r) => chainIds.has(r.sourceId));
      if (chainSuccessors.length === 0) {
        leafNodes.push(minion);
      }
    }

    // Return the leaf with the latest createdAt
    leafNodes.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
    return leafNodes[0]!;
  }

  /**
   * Returns the version that was active at a specific point in time.
   *
   * Returns the most recently created version with a `createdAt` ≤ the given date.
   *
   * @param promptId - The ID of any minion in the version chain.
   * @param date - The target date/time.
   * @returns The version active at that date, or null if none existed yet.
   */
  async getVersionAtDate(promptId: string, date: Date): Promise<Minion | null> {
    const chain = await this.getVersionChain(promptId);
    const targetMs = date.getTime();

    const candidates = chain.filter(
      (m) => new Date(m.createdAt).getTime() <= targetMs,
    );

    if (candidates.length === 0) return null;

    // Return the latest one at or before the given date
    candidates.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
    return candidates[0]!;
  }

  /** Walks backwards via `follows` targets to find the chain root. */
  private async findRoot(startId: string): Promise<Minion> {
    const visited = new Set<string>();
    let currentId = startId;

    while (true) {
      if (visited.has(currentId)) {
        throw new Error(`Cycle detected in follows chain at ${currentId}`);
      }
      visited.add(currentId);

      // Find what this version follows (i.e., sourceId = currentId, type = follows)
      const followsRelations = await this.storage.getRelations({
        sourceId: currentId,
        type: 'follows',
      });

      if (followsRelations.length === 0) {
        // This is the root
        const minion = await this.storage.getMinion(currentId);
        if (!minion) throw new Error(`Minion not found: ${currentId}`);
        return minion;
      }

      // Walk to the predecessor
      currentId = followsRelations[0]!.targetId;
    }
  }
}
